contrib/x264/A02-threads-pool-adjustments.patch

```diff
@@ -0,0 +1,67 @@
+
+---
+ common/base.c | 34 ++++++++++++++++++++++++++++++++++
+ 1 file changed, 34 insertions(+)
+
+diff --git a/common/base.c b/common/base.c
+index d302fed..298f6d3 100644
+--- a/common/base.c
++++ b/common/base.c
+@@ -27,6 +27,7 @@
+ #include "base.h"
+ 
+ #include <ctype.h>
++#include<string.h>
+ 
+ #if HAVE_MALLOC_H
+ #include <malloc.h>
+@@ -338,6 +339,33 @@ REALIGN_STACK void x264_picture_clean( x264_picture_t *pic )
+     memset( pic, 0, sizeof( x264_picture_t ) );
+ }
+ 
++#ifdef _WIN32 && HAVE_AARCH64
++int checkSnapdragon()
++{
++    int BUFFER_SIZE = 128;
++    char powershell_output[BUFFER_SIZE * 10];
++    FILE* pipe = _popen("powershell -ExecutionPolicy Bypass -Command \"(Get-CimInstance -ClassName Win32_Processor).Name\" 2>&1", "r");
++
++    char buffer[BUFFER_SIZE];
++
++    while (fgets(buffer, BUFFER_SIZE, pipe) != NULL)
++    {
++        strncat(powershell_output, buffer, BUFFER_SIZE);
++    }
++
++    _pclose(pipe);
++
++    if (strstr(powershell_output, "Snapdragon"))
++    {
++        return 1;
++    }
++    else
++    {
++       return 0;
++    }
++}
++#endif
++
+ /****************************************************************************
+  * x264_param_default:
+  ****************************************************************************/
+@@ -352,6 +380,12 @@ REALIGN_STACK void x264_param_default( x264_param_t *param )
+     param->i_lookahead_threads = X264_THREADS_AUTO;
+     param->b_deterministic = 1;
+     param->i_sync_lookahead = X264_SYNC_LOOKAHEAD_AUTO;
++#ifdef _WIN32 && HAVE_AARCH64
++    if (checkSnapdragon() && (x264_cpu_num_processors() >= 10))
++    {
++        param->i_threads = 6;
```

Comment:
Why is this helping performance?

Reply:
With the default framethreads setting chosen by x264, the CPU wasn't being fully utilized. After running multiple tests with framethreads values ranging from 1 to 12, this seemed to yield the best results.

Reply:
What do you mean by "seemd to yield the best results"? Highest encode framerate? Highest cpu utilization as measured by some tool?

My guess is that by lowering the number of threads used you are actually lowering cpu core utilization and releaving some thermal pressure.  This would allow the remaining cores to run at higher clocks and provide higher encode framerates in certain senarios.

There are a number of problems with this.
- A system using the same cpu that has better thermal management (or is running in a cooler environment)  will perform better with more threads.
- Different video source characteristics are going to result in different results. Higher resolutions usually benefit more from higher thread counts. You may find, with additional testing, that this change does not provide the best results for all types of video.
- This parameter is not the only thing that determines the number of threads that will be used in an encode. Other filters will also add more threads.

Given the state of thermally aware schedulers in the OS, I can understand how this kind of thing can have some benefits. But I don't think HandBrake has access to enough information to provide the best results. A fixed number here isn't going to provide best results for everyone. A user tunable parameter that scales the number of threads (and gets applied everywhere HandBrake spawns threads) would be an option. But most users wouldn't know how to do appropriate performance tests and apply such and option. IMO this is wondering into territory that is beyond HandBrake's primary scope.